src/v/serde/parquet/writer.cc

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "serde/parquet/writer.h"
+
+#include "bytes/iobuf.h"
+#include "bytes/iostream.h"
+#include "container/contiguous_range_map.h"
+#include "serde/parquet/column_writer.h"
+#include "serde/parquet/metadata.h"
+#include "serde/parquet/shredder.h"
+
+namespace serde::parquet {
+
+class writer::impl {
+public:
+    impl(options opts, ss::output_stream<char> output)
+      : _opts(std::move(opts))
+      , _output(std::move(output)) {}
+
+    ss::future<> init() {
+        index_schema(_opts.schema);
+        _opts.schema.for_each([this](const schema_element& element) {
+            if (!element.is_leaf()) {
+                return;
+            }
+            _column_writers.emplace(element.position, element);
+            _schema_leaves.emplace(element.position, &element);
+        });
+        // write the leading magic bytes
+        co_await write_iobuf(iobuf::from("PAR1"));
+    }
+
+    ss::future<> write_row(group_value row) {
+        co_await shred_record(
+          _opts.schema, std::move(row), [this](shredded_value sv) {
+              return write_value(std::move(sv));
+          });
+        // TODO: periodically flush the row_group if we're using enough memory
+        // or at least flush the pages if we don't want to create too many row
+        // groups.
+    }
+
+    ss::future<> close() {
+        chunked_vector<row_group> row_groups;
+        row_groups.push_back(co_await flush_row_group());
+        int64_t num_rows = 0;
+        for (const auto& rg : row_groups) {
+            num_rows += rg.num_rows;
```

Comment:
we could probably keep a running count of num_rows (which is probably needed anyway once we periodically flush).

Reply:
Yes probably. For now I'll punt on that - stats are going to get a bit complex and I'd like to keep them isolated so it doesn't distract the reader from what's going on.

Reply:
To clarify - I think it's going to be better if long term there is a separate layer for computing stats, there are a lot of things to track. That should be trivially testable (compared to this you really need a reader to test properly).